### PR TITLE
retain type of `_modules` when renaming keys

### LINKED ISSUE #2789: Changing _modules dict type in Pytorch 2.5 leads to failing collections test
## 🐛 Bug

As of PyTorch 2.5, the `Module._modules` attribute has [changed](https://github.com/pytorch/pytorch/commit/6120aa3718737dac685eb779ccc4539d4ba059d2#diff-4be56271f7bfe650e3521c81fd363da58f109cd23ee80d243156d2d6ccda6263L497) from an `OrderedDict` to a `dict`. This was done to improve the performance of certain TorchDynamo guard mechanisms and is not expected to affect other behavior because, as of Python 3.7, dictionaries are [ordered by default](https://docs.python.org/3/whatsnew/3.7.html).

This means that [MetricCollection.keys](https://github.com/Lightning-AI/torchmetrics/blob/10339a27424c8fb38ad459082a8a8405dac8e119/src/torchmetrics/collections.py#L513) returns type `dict_keys` when `keep_base=True` but returns `odict_keys` when `keep_base=False` because it uses [_to_renamed_ordered_dict](https://github.com/Lightning-AI/torchmetrics/blob/10339a27424c8fb38ad459082a8a8405dac8e119/src/torchmetrics/collections.py#L502) to construct a new `OrderedDict` (MetricCollection.items has the same issue).

As a result, [test_metric_collection_prefix_postfix_args](https://github.com/Lightning-AI/torchmetrics/blob/6a23b38bae23930de18798cde96a927ad0029b33/tests/unittests/bases/test_collections.py#L164) fails at line 200 because it calls `.keys()` with `keep_base` equal to both True and False and asserts that the resulting types are the same, which they are not.

### Proposed Fix

Change `_to_renamed_ordered_dict` so that it produces either a `dict` or an `OrderedDict`, depending on the type of the existing `_modules` dictionary. This will work for both older and newer version of PyTorch.

Current:
```py
def _to_renamed_ordered_dict(self) -> OrderedDict:
    od = OrderedDict()
    for k, v in self._modules.items():
        od[self._set_name(k)] = v
    return od
```

Proposed:
```py
def _to_renamed_dict(self) -> dict:
    if isinstance(self._modules, OrderedDict):
        d = OrderedDict()
    else:
        d = dict()
    for k, v in self._modules.items():
        d[self._set_name(k)] = v
    return d
```

### To Reproduce

Run the unit tests with PyTorch 2.5

### PR DESCRIPTION
**Pull request by bfolie**
## What does this PR do?

Fixes #2789 

<details>
  <summary>Before submitting</summary>

- [X] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [X] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
📚 Documentation preview 📚: https://torchmetrics--2793.org.readthedocs.build/en/2793/

### CODE REVIEW

File: src/torchmetrics/collections.py [hunk 1]
```diff
@@ -499,11 +499,12 @@ def _set_name(self, base: str) -> str:
         name = base if self.prefix is None else self.prefix + base
         return name if self.postfix is None else name + self.postfix
 
-    def _to_renamed_ordered_dict(self) -> OrderedDict:
-        od = OrderedDict()
+    def _to_renamed_dict(self) -> dict:
+        # self._modules changed from OrderedDict to dict as of PyTorch 2.5.0
+        d = OrderedDict() if isinstance(self._modules, OrderedDict) else dict
```
Comment (Borda) [inline src/torchmetrics/collections.py:504]:
```suggestion
        dict_modules = OrderedDict() if isinstance(self._modules, OrderedDict) else dict
```

File: src/torchmetrics/collections.py [hunk 2]
```diff
@@ -499,11 +499,12 @@ def _set_name(self, base: str) -> str:
...
         for k, v in self._modules.items():
-            od[self._set_name(k)] = v
-        return od
+            d[self._set_name(k)] = v
+        return d
```
Comment (Borda) [inline src/torchmetrics/collections.py:507]:
```suggestion
            dict_modules[self._set_name(k)] = v
        return dict_modules
```

**Review by Borda — APPROVED** [review]

File: src/torchmetrics/collections.py [hunk 3]
```diff
@@ -499,11 +499,12 @@ def _set_name(self, base: str) -> str:
         name = base if self.prefix is None else self.prefix + base
         return name if self.postfix is None else name + self.postfix
 
-    def _to_renamed_ordered_dict(self) -> OrderedDict:
-        od = OrderedDict()
+    def _to_renamed_dict(self) -> dict:
```
Comment (Borda) [inline src/torchmetrics/collections.py:502]:
```suggestion
    def _to_renamed_dict(self) -> Mapping:
```

File: src/torchmetrics/collections.py [hunk 4]
```diff
@@ -499,11 +499,12 @@ def _set_name(self, base: str) -> str:
         name = base if self.prefix is None else self.prefix + base
         return name if self.postfix is None else name + self.postfix
 
-    def _to_renamed_ordered_dict(self) -> OrderedDict:
-        od = OrderedDict()
+    def _to_renamed_dict(self) -> Mapping[str, Metric]:
+        # self._modules changed from OrderedDict to dict as of PyTorch 2.5.0
+        dict_modules = OrderedDict() if isinstance(self._modules, OrderedDict) else dict()
```
Comment (Borda) [inline src/torchmetrics/collections.py:504]:
```suggestion
        dict_modules = OrderedDict() if isinstance(self._modules, OrderedDict) else {}
```

**Comment by Borda** [comment]
@SkafteNicki one issue remained:
```
FAILED unittests/wrappers/test_multitask.py::test_error_on_wrong_keys - AssertionError: Regex pattern did not match.
 Regex: "Expected\\ arguments\\ `task_preds`\\ and\\ `task_targets`\\ to\\ have\\ the\\ same\\ keys\\ as\\ the\\ wrapped\\ `task_metrics`\\.\\ Found\\ task_preds\\.keys\\(\\)\\ =\\ dict_keys\\(\\['Classification'\\]\\),\\ task_targets\\.keys\\(\\)\\ =\\ dict_keys\\(\\['Classification',\\ 'Regression'\\]\\)\\ and\\ self\\.task_metrics\\.keys\\(\\)\\ =\\ odict_keys\\(\\['Classification',\\ 'Regression'\\]\\)"
 Input: "Expected arguments `task_preds` and `task_targets` to have the same keys as the wrapped `task_metrics`. Found task_preds.keys() = dict_keys(['Classification']), task_targets.keys() = dict_keys(['Classification', 'Regression']) and self.task_metrics.keys() = dict_keys(['Classification', 'Regression'])"
```

**Review by SkafteNicki — APPROVED** [review]

**Pull request merged**